File: backend/models/voucher.py
from sqlalchemy import Column, Integer, String, Float, DateTime, Boolean, JSON
from sqlalchemy.sql import func
from pydantic import BaseModel
from typing import List, Optional
from models.database import Base

# SQLAlchemy 모델
class Voucher(Base):
    __tablename__ = "vouchers"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(50), nullable=False)
    type = Column(String(20), nullable=False)  # pro, basic, light
    price = Column(Float, nullable=False)
    period = Column(String(20), nullable=False)  # month, week
    features = Column(JSON, nullable=False)
    is_active = Column(Boolean, default=True)

class UserVoucher(Base):
    __tablename__ = "user_vouchers"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, index=True)
    voucher_id = Column(Integer, index=True)
    is_active = Column(Boolean, default=True)
    purchased_at = Column(DateTime, default=func.now())
    expires_at = Column(DateTime, nullable=True)

class PaymentHistory(Base):
    __tablename__ = "payment_history"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, index=True)
    voucher_id = Column(Integer, index=True)
    amount = Column(Float, nullable=False)
    status = Column(String(20), nullable=False)  # completed, cancelled, pending, failed
    payment_method = Column(String(50), nullable=True)
    created_at = Column(DateTime, default=func.now())

class PaymentMethod(Base):
    __tablename__ = "payment_methods"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, index=True)
    card_type = Column(String(50), nullable=False)
    card_number = Column(String(20), nullable=False)
    expiry_date = Column(String(10), nullable=False)
    is_default = Column(Boolean, default=False)

# Pydantic 스키마
class VoucherBase(BaseModel):
    name: str
    type: str
    price: float
    period: str
    features: List[str]
    is_active: bool = True

class VoucherCreate(VoucherBase):
    pass

class VoucherResponse(VoucherBase):
    id: int

    class Config:
        from_attributes = True

class UserVoucherBase(BaseModel):
    user_id: int
    voucher_id: int
    is_active: bool = True

class UserVoucherCreate(UserVoucherBase):
    pass

class UserVoucherResponse(UserVoucherBase):
    id: int
    purchased_at: str
    expires_at: Optional[str] = None

    class Config:
        from_attributes = True

class PaymentHistoryBase(BaseModel):
    user_id: int
    voucher_id: int
    amount: float
    status: str
    payment_method: Optional[str] = None

class PaymentHistoryCreate(PaymentHistoryBase):
    pass

class PaymentHistoryResponse(PaymentHistoryBase):
    id: int
    created_at: str

    class Config:
        from_attributes = True

class PaymentMethodBase(BaseModel):
    user_id: int
    card_type: str
    card_number: str
    expiry_date: str
    is_default: bool = False

class PaymentMethodCreate(PaymentMethodBase):
    pass

class PaymentMethodResponse(PaymentMethodBase):
    id: int

    class Config:
        from_attributes = True 

File: frontend/lib/models/voucher.dart
class Voucher {
  final String id;
  final String name;
  final String type;
  final double price;
  final String period;
  final List<String> features;
  final bool isActive;

  Voucher({
    required this.id,
    required this.name,
    required this.type,
    required this.price,
    required this.period,
    required this.features,
    required this.isActive,
  });

  factory Voucher.fromJson(Map<String, dynamic> json) {
    return Voucher(
      id: json['id'].toString(),
      name: json['name'],
      type: json['type'],
      price: json['price'].toDouble(),
      period: json['period'],
      features: List<String>.from(json['features']),
      isActive: json['is_active'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'type': type,
      'price': price,
      'period': period,
      'features': features,
      'is_active': isActive,
    };
  }
}

class PaymentHistory {
  final String id;
  final String voucherName;
  final double amount;
  final DateTime date;
  final String status;

  PaymentHistory({
    required this.id,
    required this.voucherName,
    required this.amount,
    required this.date,
    required this.status,
  });

  factory PaymentHistory.fromJson(Map<String, dynamic> json) {
    return PaymentHistory(
      id: json['id'].toString(),
      voucherName: json['voucher_name'],
      amount: json['amount'].toDouble(),
      date: DateTime.parse(json['date']),
      status: json['status'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'voucher_name': voucherName,
      'amount': amount,
      'date': date.toIso8601String(),
      'status': status,
    };
  }
}

class PaymentMethod {
  final String id;
  final String cardType;
  final String cardNumber;
  final String expiryDate;

  PaymentMethod({
    required this.id,
    required this.cardType,
    required this.cardNumber,
    required this.expiryDate,
  });

  factory PaymentMethod.fromJson(Map<String, dynamic> json) {
    return PaymentMethod(
      id: json['id'].toString(),
      cardType: json['card_type'],
      cardNumber: json['card_number'],
      expiryDate: json['expiry_date'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'card_type': cardType,
      'card_number': cardNumber,
      'expiry_date': expiryDate,
    };
  }
} 